### PR TITLE
login to cluster and configure cli

### PR DESCRIPTION
**Pull request by falseperfection**
Currently CLI configuration is left to the user or whatever script is calling tests. This is an essential part of running the test and is therefore now captures in test setup

### CODE REVIEW

File: testing/sdk_login.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+def configure_cli(dcosurl: str, token: str) -> None:
+    out, err, rc = shakedown.run_dcos_command('dcos config set core.dcos_url {}'.format(dcosurl))
+    assert rc, 'Failed to set core.dcos_url: {}'.format(err)
+    out, err, rc = shakedown.run_dcos_command('dcos config set core.ssl_verify false')
+    assert rc, 'Failed to set core.ssl_verify: {}'.format(err)
+    out, err, rc = shakedown.run_dcos_command('config set core.dcos_acs_token {}'.format(token))
+    assert rc, 'Failed to set core.dcos_acs_token: {}'.format(err)
+
+
+def logout(dcosurl: str):
```
Comment (mellenburg) [inline testing/sdk_login.py]: Do we even care? Seems OK to just leave off now and forever

File: testing/sdk_login.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import logging
+import os
+
+import requests
+import shakedown
```
Comment (mellenburg) [inline testing/sdk_login.py]: I would prefer if we do not use shakedown in this case. See below

File: testing/sdk_login.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    headers = {'Content-Type': 'application/json',}
+    login_endpoint = '{dcosurl}/acs/api/v1/auth/login'.format(dcosurl=dcosurl)
+    r = requests.post(login_endpoint, headers=headers, json=payload, verify=False)
+    assert r.status_code == 200, '{} failed {}: {}'.format(login_endpoint, r.status_code, r.text)
+
+    return r.json()['token']
+
+
+def login(dcosurl: str, username: str, password: str, is_enterprise: bool, user_token: str) -> str:
```
Comment (mellenburg) [inline testing/sdk_login.py]: Why bother passing user_token to this function at all if it just gets passed through untouched?

File: testing/sdk_login.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        dcos_login_username = os.environ.get('DCOS_LOGIN_USERNAME', __CLI_LOGIN_EE_USERNAME)
+        dcos_login_password = os.environ.get('DCOS_LOGIN_PASSWORD', __CLI_LOGIN_EE_PASSWORD)
+        dcos_enterprise = os.environ.get('DCOS_ENTERPRISE', False) and True
+        dcos_acs_token = os.environ.get('DCOS_ACS_TOKEN')
+        token = login(
+            dcosurl=cluster_url,
+            username=dcos_login_username,
+            password=dcos_login_password,
+            is_enterprise=dcos_enterprise,
+            user_token=dcos_acs_token
```
Comment (mellenburg) [inline testing/sdk_login.py]: Rather than have the login function use the `user_token` argument, I would prefer that we move the if statement out of that function to here. E.G.:
```
if not dcos_acs_token:
    dcos_acs_token = login(...)
configure_cli(dcos_url, dcos_acs_token)
```
It makes a little more sense to me given that the dcos_acs_token can never really be used in the login process. It just negates the need to login at all. WDYT?

File: testing/sdk_login.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    return user_token
+
+
+def configure_cli(dcosurl: str, token: str) -> None:
+    out, err, rc = shakedown.run_dcos_command('dcos config set core.dcos_url {}'.format(dcosurl))
+    assert rc, 'Failed to set core.dcos_url: {}'.format(err)
+    out, err, rc = shakedown.run_dcos_command('dcos config set core.ssl_verify false')
+    assert rc, 'Failed to set core.ssl_verify: {}'.format(err)
+    out, err, rc = shakedown.run_dcos_command('config set core.dcos_acs_token {}'.format(token))
+    assert rc, 'Failed to set core.dcos_acs_token: {}'.format(err)
```
Comment (mellenburg) [inline testing/sdk_login.py]: can we remove the shakedown call from this? in this case here, the function is literally equivalent to:
```
import subprocess

def run_dcos_command(cmd_str):
    subprocess.check_call(['dcos'] + cmd_str.split())

def configure_cli(dcosurl: str, token: str) -> None:
    run_dcos_command('config set core.dcos_url {}'.format(dcosurl))
    run_dcos_command('config set core.ssl_verify false')
    run_dcos_command('config set core.dcos_acs_token {}'.format(token))
```
That will automatically raise an assertion that contains the output of the failed command. If you want to embellish the errors then you can do so in run_dcos_command with:
```
try:
    # do stuff in run_dcos_command
except subprocess.CalledProcessError as ex:
    raise Exception('Failed to {}: {}'.format(cmd_str, ex.stdout.decode()))
```
I don't think its really necessary (i.e. the former is enough of an implementation), but I leave that to you to decide

Comment (mellenburg) [inline testing/sdk_login.py]: Also, I kept the command input as a string for consistency with the current example, but I think it would be best to pass through a list to subprocess.check_call. Thats only a small preference though. The important thing is that the command is passed as a list to `check_call` as passing a str is platform dependent in execution

File: testing/sdk_login.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        dcos_acs_token = os.environ.get('DCOS_ACS_TOKEN')
+        token = login(
+            dcosurl=cluster_url,
+            username=dcos_login_username,
+            password=dcos_login_password,
+            is_enterprise=dcos_enterprise,
+            user_token=dcos_acs_token
+        )
+        configure_cli(dcosurl, token)
+        yield
```
Comment (mellenburg) [inline testing/sdk_login.py]: Not required at all, but what do you think about yielding the acs token here? That way any test that needs it can just take it from this fixture instead of having to make a call to retrieve it with shakedown, which in turn calls out to the dcoscli

File: testing/sdk_login.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    finally:
+        logout(dcosurl=cluster_url)
```
Comment (mellenburg) [inline testing/sdk_login.py]: Unless you see a clear benefit to logging out, I would be fine with dropping the whole `try:finally` clause here

File: frameworks/cassandra/tests/conftest.py [hunk 8]
```diff
@@ -1,10 +1,15 @@
 import pytest
+import sdk_login
 import sdk_repository
 import sdk_security
 
 
 @pytest.fixture(scope='session')
-def configure_universe():
+def configure_login():
```
Comment (mellenburg) [inline frameworks/cassandra/tests/conftest.py]: This invocation is identical in all of the framework test suites so I am fine if you make it an implicit (i.e. autouse) or explicit fixture at the top-level and thereby avoid having to make all these small changes.

One advantage of doing this at the top-level conftest.py with implicit injection is that you could cd down to this individual directory and login yourself and not deal with this new interface or simply avoid having the test reconfigure your local CLI

However, that's a pretty specific scenario and stuff at the top level might just need to be copied out anyway if we split these tests into their own repos. Furthermore the function you wrote already has a clear override mechanism so its not strictly necessary.

File: testing/sdk_login.py (hunk 3 above)
Comment (falseperfection) [inline testing/sdk_login.py]: Because even people as perfect as me do dumb things sometimes. :)

File: testing/sdk_login.py (hunk 5 above)
Comment (falseperfection) [inline testing/sdk_login.py]: I prefer consistency over correctness most of the time. There are `shakedown.run_docs_command` calls throughout. If we change here, I want to see it changed everywhere. That means we'd need a shared `run_dcos_ command`, bringing us very nearly back to where we are now. Just instead of `shakedown.run_dcos_command` it would be `sdk_utils.run_dcos_command`. I'm not sure I understand the value there. (Other than getting rid of the _word_ shakedown, but I have no issue with shakedown.)

File: testing/sdk_login.py (hunk 1 above)
Comment (falseperfection) [inline testing/sdk_login.py]: Yeah. I think I'm over thinking about cleanup. I believe it's a good practice, but I don't see us doing it anytime soon. (Nobody but me cares, and I don't have a strongly compelling use-case.)

File: testing/sdk_login.py (hunk 4 above)
Comment (falseperfection) [inline testing/sdk_login.py]: Well obviously. ;)

File: testing/sdk_login.py (hunk 5 above)
Comment (mellenburg) [inline testing/sdk_login.py]: That's a fair point and I do mind just pushing that out to another issue where we apply it consistently. I definitely have a tendency to push for "landing" paradigms over maintaining consistency...

Given that we can agree to table this, my qualm with shakedown in this case is the accessibility of the interface. Once I use shakedown for a CLI command, I am extremely limited in what I can  do with that command: https://github.com/dcos/shakedown/blob/master/shakedown/dcos/command.py#L103

By using subprocess, I have the ability to:
* declare a timeout
* arbitrarily redirect stdout, stdin, and stderr
* implicitly raise an expception on failure
* specifically control the logging statements around this command
* specify if the command should be shell interpreted
* control the specific environment for that command
* provide my command in an unambiguous argument list

By using shakedown:
* I MUST submit my command as a string that will be split up again anyway. This makes it harder to provide intricate arguments.
* I MUST push all output to either subprocess.PIPE or stdout. This prevents me from streaming the output
* I MUST declare with a keyword arg that failures in the command should raise a local exception
* If I want to access any of the powerful subprocess options ( https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module ) then I need to make a commit to shakedown and then wait for a dependency update in this repo. This means I cannot control: timeouts, working dir, command entrypoint (shell), or custom decoders for I/O (might be useful for mesos task log retrieval)

TL;DR I see the argument on maintaining consistency so we can handle this later, but I really care about making sure tests can directly access the powerful, native subprocess interface instead of using a customized, restrictive interface

File: testing/sdk_login.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+    @pytest.fixture(scope='session')
+    def configure_login():
+        yield from sdk_login.login_session()
+    """
+    try:
+        cluster_url = os.environ.get('CLUSTER_URL')
+        dcos_login_username = os.environ.get('DCOS_LOGIN_USERNAME', __CLI_LOGIN_EE_USERNAME)
+        dcos_login_password = os.environ.get('DCOS_LOGIN_PASSWORD', __CLI_LOGIN_EE_PASSWORD)
+        dcos_enterprise = os.environ.get('DCOS_ENTERPRISE', False) and True
```
Comment (mellenburg) [inline testing/sdk_login.py]: wont this evaluate true if you set 'DCOS_ENTERPRISE=false'?

Comment (mellenburg) [inline testing/sdk_login.py]: My preference for this pattern is generally:
```
dcos_enterprise = os.getenv('DCOS_ENTERPRISE', 'true') == 'true'
```
Also, I intentionally went with enterprise as the default to be consistent with how the testing suite is biased now, even though that feels like a faux pas

Comment (falseperfection) [inline testing/sdk_login.py]: Yes, it will. Because it's a boolean, the idiom I was going for was "if the variable is defined," not the value it's set to. So things like case and "1" vs "True" vs "Yes" didn't have to be worried about. Either way needs to be made more clear.

File: frameworks/cassandra/tests/conftest.py (hunk 8 above)
Comment (falseperfection) [inline frameworks/cassandra/tests/conftest.py]: I don't want to do top-level because I like thinking of the frameworks as separate entities as much as possible. If anything, I'd rather go back to how I had the setup working originally:

```
@pytest.fixture(scope='session')
def configuire_session():
  try:
    login_config = sdk_login.configure_login()
    repo_config = sdk_repository.configure_repository()
    security_config = sdk_security.configure_security('cassandra')
    yield
  finally:
    sdk_security.cleanup_security(security_config)
    sdk_repository.cleanup_repository(repo_config)
    sdk_login.cleanup_login(login_config)
```

I think that makes the setup clean, explicit, and minimizes the boilerplate, but I'm not tied to any implementation. I'll leave it as-is as a matter of inertia.

File: testing/sdk_login.py (hunk 5 above)
Comment (falseperfection) [inline testing/sdk_login.py]: Ticket created, and my next comment added there - https://jira.mesosphere.com/browse/INFINITY-2182

File: test.sh [hunk 10]
```diff
@@ -124,6 +130,7 @@ docker run --rm \
     -e CLUSTER_URL=$CLUSTER_URL \
     $azure_args \
     -e SECURITY=$security \
+    -e DCOS_ENTERPRISE=DCOS_ENTERPRISE \
```
Comment (elezar) [inline test.sh]: You will need to access the variable value here: `"$DCOS_ENTERPRISE"`

Comment (falseperfection) [inline test.sh]: 👍 Good catch

**Comment by benclarkwood** [comment]
I think this is redundant?
